File: backend/src/bond_monitor/application/trading/sandbox_use_case.py
"""Sandbox account management use cases."""

from __future__ import annotations

from bond_monitor.application.trading.context import TradingContext
from bond_monitor.domain.portfolio.models import PortfolioMode
from bond_monitor.domain.shared.money import Rub
from bond_monitor.domain.trading.models import AccountKind
from bond_monitor.application.trading import broker


def _linked_portfolio_dict(portfolio) -> dict | None:
    if portfolio is None:
        return None
    return {"id": portfolio.id, "name": portfolio.name}


class SandboxUseCase:
    def __init__(self, ctx: TradingContext) -> None:
        self._ctx = ctx

    async def list_accounts(self, kind: AccountKind) -> list[dict]:
        accounts = broker.list_accounts(self._ctx.token(kind), kind)
        linked_by_account_id = {
            portfolio.account_id: portfolio
            for portfolio in await self._ctx.repo.list_all()
            if (
                portfolio.mode == PortfolioMode.TRADING
                and portfolio.account_id
                and portfolio.account_kind == kind
            )
        }
        return [
            {
                "id": account.id,
                "name": account.name,
                "kind": kind.value,
                "linked_portfolio": _linked_portfolio_dict(
                    linked_by_account_id.get(account.id),
                ),
            }
            for account in accounts
        ]

    async def delete_sandbox_account(self, account_id: str) -> dict:
        kind = AccountKind.SANDBOX
        token = self._ctx.token(kind)
        linked = await self._ctx.find_linked_portfolio(account_id=account_id, kind=kind)
        deleted_portfolio_id: str | None = None
        if linked is not None:
            deleted_portfolio_id = linked.id
            await self._ctx.repo.delete(linked.id)
        broker.close_sandbox_account(token, account_id)
        return {
            "account_id": account_id,
            "deleted_portfolio_id": deleted_portfolio_id,
        }

    async def sandbox_pay_in_for_portfolio(
        self,
        portfolio_id: str,
        *,
        amount_rub: float,
    ) -> dict:
        portfolio = await self._ctx.get_trading_portfolio(portfolio_id)
        if portfolio.account_kind != AccountKind.SANDBOX:
            raise ValueError("Пополнение доступно только для песочницы")
        if amount_rub <= 0:
            raise ValueError("Сумма пополнения должна быть больше нуля")
        token = self._ctx.token(AccountKind.SANDBOX)
        balance = broker.sandbox_pay_in(token, portfolio.account_id, Rub(amount_rub))  # type: ignore[arg-type]
        return {
            "amount_added_rub": amount_rub,
            "money_rub": float(balance),
        }

    async def create_sandbox_account(
        self,
        *,
        initial_amount_rub: float,
        name: str | None = None,
    ) -> dict:
        if initial_amount_rub <= 0:
            raise ValueError("Сумма пополнения должна быть больше нуля")
        token = self._ctx.token(AccountKind.SANDBOX)
        account_name = (name or "bond-monitor").strip() or "bond-monitor"
        account_id = broker.open_sandbox_account(token, name=account_name)
        balance = broker.sandbox_pay_in(token, account_id, Rub(initial_amount_rub))
        return {
            "id": account_id,
            "name": account_name,
            "kind": AccountKind.SANDBOX.value,
            "money_rub": float(balance),
        }

File: backend/internal/application/trading/sandbox_use_case.go
package trading

import (
	"context"
	"fmt"

	domainPortfolio "github.com/tonatos/instrumenta/backend/internal/domain/portfolio"
	"github.com/tonatos/instrumenta/backend/internal/domain/shared"
	"github.com/tonatos/instrumenta/backend/internal/domain/trading"
	"github.com/tonatos/instrumenta/backend/internal/interfaces/auth"
)

// SandboxUseCase manages sandbox accounts.
type SandboxUseCase struct {
	ctx    *Context
	broker *BrokerFacade
}

func NewSandboxUseCase(ctx *Context, broker *BrokerFacade) *SandboxUseCase {
	return &SandboxUseCase{ctx: ctx, broker: broker}
}

func (u *SandboxUseCase) ListAccounts(ctx context.Context, kind trading.AccountKind) ([]map[string]any, error) {
	accounts, err := u.broker.ListAccounts(ctx, kind)
	if err != nil {
		return nil, err
	}
	owner, ok := auth.OwnerTelegramID(ctx)
	var all []domainPortfolio.Portfolio
	var listErr error
	if ok {
		all, listErr = u.ctx.Repo().ListByOwner(ctx, owner)
	} else {
		all, listErr = u.ctx.Repo().ListAll(ctx)
	}
	if listErr != nil {
		return nil, listErr
	}
	linked := map[string]domainPortfolio.Portfolio{}
	for _, p := range all {
		if p.Mode == domainPortfolio.PortfolioModeTrading && p.AccountID != nil && p.AccountKind != nil && *p.AccountKind == kind {
			linked[*p.AccountID] = p
		}
	}
	result := make([]map[string]any, 0, len(accounts))
	for _, account := range accounts {
		item := map[string]any{
			"id": account.ID, "name": account.Name, "kind": string(kind),
			"is_writable": account.IsWritable,
		}
		if lp, ok := linked[account.ID]; ok {
			item["linked_portfolio"] = map[string]string{"id": lp.ID, "name": lp.Name}
		}
		result = append(result, item)
	}
	return result, nil
}

func (u *SandboxUseCase) CreateSandboxAccount(ctx context.Context, initialAmountRub float64, name string) (map[string]any, error) {
	if initialAmountRub <= 0 {
		return nil, fmt.Errorf("amount must be positive")
	}
	accountName := name
	if accountName == "" {
		accountName = "instrumenta"
	}
	accountID, err := u.broker.OpenSandboxAccount(ctx, accountName)
	if err != nil {
		return nil, err
	}
	balance, err := u.broker.SandboxPayIn(ctx, accountID, shared.Rub(initialAmountRub))
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"id": accountID, "name": accountName, "kind": string(trading.AccountKindSandbox),
		"money_rub": float64(balance),
	}, nil
}

func (u *SandboxUseCase) DeleteSandboxAccount(ctx context.Context, accountID string) (map[string]any, error) {
	kind := trading.AccountKindSandbox
	linked, _ := u.ctx.FindLinkedPortfolio(ctx, accountID, kind, "")
	var deletedPortfolioID *string
	if linked != nil {
		deletedPortfolioID = &linked.ID
		if _, err := u.ctx.Repo().Delete(ctx, linked.ID); err != nil {
			return nil, err
		}
	}
	if err := u.broker.CloseSandboxAccount(ctx, accountID); err != nil {
		return nil, err
	}
	result := map[string]any{"account_id": accountID}
	if deletedPortfolioID != nil {
		result["deleted_portfolio_id"] = *deletedPortfolioID
	}
	return result, nil
}

func (u *SandboxUseCase) SandboxPayInForPortfolio(ctx context.Context, portfolioID string, amountRub float64) (map[string]any, error) {
	p, err := u.ctx.GetTradingPortfolio(ctx, portfolioID)
	if err != nil {
		return nil, mapTradingErr(err)
	}
	if p.AccountKind == nil || *p.AccountKind != trading.AccountKindSandbox {
		return nil, fmt.Errorf("Пополнение доступно только для песочницы")
	}
	if amountRub <= 0 {
		return nil, fmt.Errorf("Сумма пополнения должна быть больше нуля")
	}
	balance, err := u.broker.SandboxPayIn(ctx, *p.AccountID, shared.Rub(amountRub))
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"amount_added_rub": amountRub,
		"money_rub":        float64(balance),
	}, nil
}
